### PR TITLE
Group Management UI

### PR DESCRIPTION
**Pull request by tilaks26**
- Added a Manage Page which lists the owner groups and member groups of the user.
- Added a Create Group button.
- Users can view the group details, delete group, leave group.

Functionalities that are yet to be implemented - 
Adding or removing group members with a user search (will be done by Monday).

Please let me know if there is anything else that is needed.

### CODE REVIEW

**Comment by tilaks26** [comment]
- Added an edit view for groups where you can add/remove members.
- Display confirmation alert message when trying to delete group/leave group/remove members from group.
- Display success messages after performing operations.

File: django_airavata/apps/groups/forms.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Create your forms here.
+from django import forms
+
+class CreateForm(forms.Form):
+    domain_id = forms.CharField(required=True, disabled=True)
```
Comment (machristie) [inline django_airavata/apps/groups/forms.py]: All of these disabled fields we should just hide. I think you can do this by setting `widget=forms.HiddenInput`

File: django_airavata/apps/groups/templates/django_airavata_groups/group_details.html [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                <table class="table table-striped table-bordered">
+                    <thead>
+                        <tr>
+                            <th style="text-align:center">PROPERTY</th>
+                            <th style="text-align:center">VALUE</th>
+                        </tr>
+                    </thead>
+                    <tbody align="center">
+                        <tr>
+                            <td>Group ID</td>
```
Comment (machristie) [inline django_airavata/apps/groups/templates/django_airavata_groups/group_details.html]: Let's hide Group ID, domain ID, group type and cardinality

File: django_airavata/apps/groups/templates/django_airavata_groups/group_edit.html [hunk 3]
```diff
@@ -0,0 +1,38 @@
+{% extends 'base.html' %}
+
+{% block content %}
+
+<h1 align="center">Edit Group</h1>
+<br>
+{% block body %}
+<div class="container">
+    {% if user.is_authenticated %}
+    <h5>Group ID: {{ group_id }}</h5>
```
Comment (machristie) [inline django_airavata/apps/groups/templates/django_airavata_groups/group_edit.html]: Instead of group_id, let's display the group name

File: django_airavata/apps/groups/templates/django_airavata_groups/groups_manage.html [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                        <th style="text-align:center">GROUP_ID</th>
+                        <th style="text-align:center">NAME</th>
+                        <th style="text-align:center">DOMAIN_ID</th>
+                        <th style="text-align:center">OPTIONS</th>
+                    </tr>
+                </thead>
+                <tbody align="center">
+                    {% for group in owner %}
+                    <tr>
+                        <td>{{ group.groupId }}</td>
```
Comment (machristie) [inline django_airavata/apps/groups/templates/django_airavata_groups/groups_manage.html]: remove group id and domain id

File: django_airavata/apps/groups/templates/django_airavata_groups/groups_manage.html [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+                        <td>{{ group.name }}</td>
+                        <td>{{ group.domainId }}</td>
+                        <td>
+                            <a href="{% url 'django_airavata_groups:view' %}?group_id={{ group.groupId }}">
```
Comment (machristie) [inline django_airavata/apps/groups/templates/django_airavata_groups/groups_manage.html]: Need to URL encode the groupId, also on lines 52 and 55. You should be able to do `{{ group.groupId | urlencode }}`.

File: django_airavata/apps/groups/templates/django_airavata_groups/groups_manage.html [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    <br>
+    <div class="row">
+        <div class="col-md-12 table-responsive">
+            <h3>Groups you're a member of</h3>
+            <br>
+            <br>
+            <table class="table table-striped table-bordered">
+                <thead>
+                    <tr>
+                        <th style="text-align:center">GROUP_ID</th>
```
Comment (machristie) [inline django_airavata/apps/groups/templates/django_airavata_groups/groups_manage.html]: Remove GROUP_ID and DOMAIN_ID

File: django_airavata/apps/groups/templates/django_airavata_groups/groups_manage.html [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                    </tr>
+                </thead>
+                <tbody align="center">
+                    {% for group in member %}
+                    <tr>
+                        <td>{{ group.groupId }}</td>
+                        <td>{{ group.name }}</td>
+                        <td>{{ group.domainId }}</td>
+                        <td>
+                            <a href="{% url 'django_airavata_groups:view' %}?group_id={{ group.groupId }}">
```
Comment (machristie) [inline django_airavata/apps/groups/templates/django_airavata_groups/groups_manage.html]: Same thing about urlencode

File: django_airavata/apps/groups/views.py [hunk 8]
```diff
@@ -2,22 +2,174 @@
...
 logger = logging.getLogger(__name__)
 
+@login_required
+def groups_manage(request):
+
+    gateway_id = settings.GATEWAY_ID
+    username = request.user.username
+
+    try:
+        owner_list = request.sharing_client.getGroups(gateway_id, 0, -1)
```
Comment (machristie) [inline django_airavata/apps/groups/views.py:23]: Don't we want to filter out the user's single user group? That is, this group management UI should only manage multi-user groups so I think we should always filter out single user groups.

File: django_airavata/apps/groups/views.py [hunk 9]
```diff
@@ -2,22 +2,174 @@
...
 @login_required
 def groups_list(request):
 
     gateway_id = settings.GATEWAY_ID
 
     try:
-        groups = request.sharing_client.getGroups(gateway_id, 0, -1)
+        group_list = request.sharing_client.getGroups(gateway_id, 0, -1)
+        for group in group_list:
+            if group.groupCardinality == 0:
```
Comment (machristie) [inline django_airavata/apps/groups/views.py]: Looks like you are removing single user groups here but I don't think the app uses groups_list any longer. Maybe remove the groups_list function but move this groupCardinality check to the groups_manage function.

File: django_airavata/apps/groups/views.py [hunk 10]
```diff
@@ -2,22 +2,174 @@
...
+    #If role = user
+    group_type = 1
+    #Else if role = admin
+    #group_type = 2
+    group_cardinality = GroupCardinality.MULTI_USER
+
+    if request.method == 'POST':
+        form = CreateForm(request.POST, request.FILES, initial={'domain_id': gateway_id, 'group_owner': username, 'group_type': group_type, 'group_cardinality': group_cardinality})
+        if form.is_valid():
+            group_id = str(uuid.uuid4().hex) + form.cleaned_data.get('group_name')
```
Comment (machristie) [inline django_airavata/apps/groups/views.py]: Put the group name first, then the UUID. Also can you replace any spaces in the group_name with hyphens when constructing the group id? You could do something similar like what you did for making the gateway id from the gateway name.

File: django_airavata/templates/base.html [hunk 11]
```diff
@@ -7,7 +7,7 @@
     <meta http-equiv="X-UA-Compatible" content="IE=edge">
     <meta name="viewport" content="width=device-width, initial-scale=1">
     <!-- The above 3 meta tags *must* come first in the head; any other head content must come *after* these tags -->
-    <title>{% block title %}Django Airavata Gateway{% endblock %}</title>
```
Comment (machristie) [inline django_airavata/templates/base.html]: Can you undo the changes to this file? We do need something like this but I need to think about how to do this for the CIPRES sharing portal.

File: django_airavata/templates/django_airavata/home.html [hunk 12]
```diff
@@ -4,11 +4,10 @@
 <div class="jumbotron">
```
Comment (machristie) [inline django_airavata/templates/django_airavata/home.html]: Likewise can you undo the changes to this file.

**Review by machristie — CHANGES_REQUESTED** [review]
Looks really good Sneha! I just have a few suggested changes. Let me know if you have any questions.

Thanks.

File: django_airavata/apps/groups/forms.py (hunk 1 above)
Comment (tilaks26) [reply django_airavata/apps/groups/forms.py]: How can I mention the widget keyword twice? I mean, we need _widget=forms.RadioSelect_ also, right?

File: django_airavata/apps/groups/templates/django_airavata_groups/group_details.html (hunk 2 above)
Comment (tilaks26) [reply django_airavata/apps/groups/templates/django_airavata_groups/group_details.html]: Hide as in, not display them?

File: django_airavata/apps/groups/forms.py (hunk 1 above)
Comment (machristie) [reply django_airavata/apps/groups/forms.py]: No, for the group_type and group_cardinality, just make the widget=forms.HiddenInput. You also don't need to specify choices for them.

The widget affects how the form is rendered. With HiddenInput it will render the form field as
```
<input type="hidden" name="..." value="...">
```

File: django_airavata/apps/groups/templates/django_airavata_groups/group_details.html (hunk 2 above)
Comment (machristie) [reply django_airavata/apps/groups/templates/django_airavata_groups/group_details.html]: That's right, I didn't really mean hide, I mean just to not display them at all.

**Review by machristie — APPROVED** [review]
